File: Battleship.Shared/Board.cs
namespace Battleship.Shared;
public class Board
{
    private readonly List<Ship> _ships = new();
    private readonly HashSet<Coordinate> _missedShots = new();
    private readonly Dictionary<Coordinate, ShotResult> _shotHistory = new();
    public IReadOnlyCollection<Ship> Ships => _ships.AsReadOnly();
    public IReadOnlyCollection<Coordinate> MissedShots => _missedShots.AsReadOnly();
    public bool CanPlaceShip(ShipType type, Coordinate start, Orientation orientation)
    {
        var newShip = new Ship(type, start, orientation);
        if (!newShip.OccupiedCoordinates.All(c => c.IsValid))
        {
            return false;
        }
        foreach (var ship in _ships)
        {
            if (newShip.OccupiedCoordinates.Any(ship.OccupiedCoordinates.Contains))
            {
                return false;
            }
        }
        return true;
    }
    public bool PlaceShip(ShipType type, Coordinate start, Orientation orientation)
    {
        if (!CanPlaceShip(type, start, orientation))
        {
            return false;
        }
        _ships.Add(new Ship(type, start, orientation));
        return true;
    }
    public ShotResult FireShot(Coordinate target)
    {
        if (!target.IsValid || _shotHistory.ContainsKey(target))
        {
            return ShotResult.Miss;
        }
        foreach (var ship in _ships)
        {
            var result = ship.FireShot(target);
            if (result != ShotResult.Miss)
            {
                _shotHistory[target] = result;
                return result;
            }
        }
        _missedShots.Add(target);
        _shotHistory[target] = ShotResult.Miss;
        return ShotResult.Miss;
    }
    public bool AreAllShipsSunk() => _ships.Any() && _ships.All(ship => ship.IsSunk);
}

File: battleships-go/internal/network/board.go
package network

import "battleships-go/pkg/models"

type Board struct {
	Ships       []*models.Ship
	MissedShots map[models.Coordinate]struct{}
	ShotHistory map[models.Coordinate]models.ShotResult
}

func NewBoard() *Board {
	return &Board{

		Ships:       make([]*models.Ship, 0),
		MissedShots: make(map[models.Coordinate]struct{}),
		ShotHistory: make(map[models.Coordinate]models.ShotResult),
	}
}

func (b *Board) CanPlaceShip(shipType models.ShipType, start models.Coordinate, orientation models.Orientation) bool {
	newShip := models.NewShip(shipType, start, orientation)
	for _, c := range newShip.OccupiedCoordinates {
		if !c.IsValid() {
			return false
		}
	}

	for _, existingShip := range b.Ships {
		for _, newCoord := range newShip.OccupiedCoordinates {
			if existingShip.ContainsCoordinate(newCoord) {
				return false
			}
		}
	}

	return true
}

func (b *Board) PlaceShip(shipType models.ShipType, start models.Coordinate, orientation models.Orientation) bool {
	if !b.CanPlaceShip(shipType, start, orientation) {
		return false
	}

	b.Ships = append(b.Ships, models.NewShip(shipType, start, orientation))
	return true
}

func (b *Board) FireShot(target models.Coordinate) models.ShotResult {
	result, _ := b.FireShotWithType(target)
	return result
}

func (b *Board) FireShotWithType(target models.Coordinate) (models.ShotResult, *models.ShipType) {
	if !target.IsValid() {
		return models.Miss, nil
	}
	if _, alreadyShot := b.ShotHistory[target]; alreadyShot {
		return models.Miss, nil
	}

	for _, ship := range b.Ships {
		result := ship.FireShot(target)

		if result != models.Miss {
			b.ShotHistory[target] = result
			if result == models.Sunk {
				st := ship.Type
				return result, &st
			}
			return result, nil
		}
	}

	b.MissedShots[target] = struct{}{}
	b.ShotHistory[target] = models.Miss

	return models.Miss, nil
}

func (b *Board) AreAllShipsSunk() bool {
	if len(b.Ships) == 0 {
		return false
	}

	for _, ship := range b.Ships {
		if !ship.IsSunk() {
			return false
		}
	}

	return true
}
